File: Recover/ios/App/App/WidgetPlugin.swift
import Foundation
import Capacitor
import WidgetKit

/**
 * Capacitor Plugin for Widget Communication
 *
 * Handles data transfer between the main app and widget extension via App Groups
 */
@objc(WidgetPlugin)
public class WidgetPlugin: CAPPlugin {

    // App Group identifier - must match in both app and widget extension
    private let appGroupIdentifier = "group.com.recover.app"
    private let widgetDataKey = "widgetData"

    @objc func updateWidgetData(_ call: CAPPluginCall) {
        guard let dataString = call.getString("data") else {
            call.reject("Missing data parameter")
            return
        }

        // Write to shared UserDefaults (App Group)
        if let userDefaults = UserDefaults(suiteName: appGroupIdentifier) {
            userDefaults.set(dataString, forKey: widgetDataKey)
            userDefaults.synchronize()

            call.resolve(["success": true])
        } else {
            call.reject("Failed to access App Group")
        }
    }

    @objc func reloadWidgets(_ call: CAPPluginCall) {
        // Reload all widgets of this app
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
            call.resolve(["success": true])
        } else {
            call.reject("Widgets require iOS 14+")
        }
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            call.resolve(["available": true])
        } else {
            call.resolve(["available": false])
        }
    }
}

File: Recover/android/app/src/main/java/com/recover/app/WidgetPlugin.kt
package com.recover.app

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

/**
 * Capacitor Plugin for Widget Communication (Android)
 *
 * Handles data transfer between the main app and widget via SharedPreferences
 */
@CapacitorPlugin(name = "WidgetPlugin")
class WidgetPlugin : Plugin() {

    private val PREFS_NAME = "com.recover.app.widget"
    private val WIDGET_DATA_KEY = "widgetData"

    @PluginMethod
    fun updateWidgetData(call: PluginCall) {
        val dataString = call.getString("data")

        if (dataString == null) {
            call.reject("Missing data parameter")
            return
        }

        try {
            // Write to SharedPreferences
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            prefs.edit().putString(WIDGET_DATA_KEY, dataString).apply()

            call.resolve(
                com.getcapacitor.JSObject().put("success", true)
            )
        } catch (e: Exception) {
            call.reject("Failed to save widget data: ${e.message}")
        }
    }

    @PluginMethod
    fun reloadWidgets(call: PluginCall) {
        try {
            // Trigger widget update
            val intent = Intent(context, RecoverWidgetProvider::class.java)
            intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE

            val appWidgetManager = AppWidgetManager.getInstance(context)
            val widgetComponent = ComponentName(context, RecoverWidgetProvider::class.java)
            val widgetIds = appWidgetManager.getAppWidgetIds(widgetComponent)

            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, widgetIds)
            context.sendBroadcast(intent)

            call.resolve(
                com.getcapacitor.JSObject().put("success", true)
            )
        } catch (e: Exception) {
            call.reject("Failed to reload widgets: ${e.message}")
        }
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        call.resolve(
            com.getcapacitor.JSObject().put("available", true)
        )
    }
}
